crates/core/src/host/module_host.rs

```
@@ -570,70 +579,26 @@ impl ModuleHost {
             )
             .await
             .map(drop)
+            .inspect_err(|_| {
+                // ...but it has failed, so we treat this as if the reducer didn't exist
+                // and commit a separate transaction instead.
```

I am not convinced this is correct in the connected path. If the reducer doesn't exist, we should run the fallback, but if the reducer exists and failed, we should return an error from this method, which IIRC will cause the connection to be rejected/terminated.

This is the intended behavior in the disconnected path, though.

---

But in both cases the client ends up disconnected, so we should have the row deleted, no?

---

The enclosing method, `call_identity_connected_disconnected`, is called in two contexts: when a new client establishes a connection (the "connected path") and when a previously-connected client terminates its connection (the "disconnected path"). These paths have meaningfully different behavior, and possibly should just be different methods.

### In the "connected path," for a new client:

- The row is expected to previously have not been present in `st_client`.
- If the module does not define a `client_connected` reducer, we should insert a new row into `st_client` and return successfully.
- If the module does define a `client_connected` reducer, call it.
  - If that reducer call returns successfully, we should insert a new row into `st_client` *within the same transaction*.
    - This is handled elsewhere, in `module_host_actor.rs`'s `WasmModuleInstance::call_reducer_with_tx`.
  - If that reducer call returns an error, we should not insert into `st_client`. Instead, we should return an error from `call_identity_connected_disconnected`, which will cause `client_connection.rs`'s `ClientConnection::spawn` to terminate the connection.
    - **This is the part that I believe to be incorrect in this PR.**
  
### In the "disconnected path," for a previously-existing client:

- The row is expected to be present in `st_client` prior to entering `call_identity_connected_disconnected`.
- If the module does not define a `client_disconnected` reducer, we should delete that row from `st_client` and return successfully.
- If the module does define a `client_disconnected` reducer, call it.
  - If that reducer call returns successfully, we should delete the row from `st_client` *within the same transaction*.
    - If done in a separate transaction, we may crash between the `client_disconnected` reducer call and the `st_client` delete, after which we will restart into a state where the reducer has already been called but the row has not been deleted, and will then re-run `client_disconnected` and re-delete the row.
    - As above, this is handled by `WasmModuleInstance::call_reducer_with_tx`.
    - **This is the part that I believe to have been broken prior to this PR, and to be fixed here.**
  - If that reducer call returns an error, we should still delete the row from `st_client`.
    - The errored reducer call's transaction is already destroyed, so we must start a new transaction to do this delete.
    - That's ok, though, since if we crash in between and restart, we'll re-run `client_disconnected`, have it fail again, and no one will ever be the wiser.